### ISSUE TITLE
rbserver - Calling `ZivoeITO.migrateDeposits`, `ZivoeRewards.getRewards`, and `ZivoeRewardsVesting.getRewards` functions can revert if `transfer` function of one of relevant tokens fails due to contract pausing or sender/receiver blocking

### ISSUE THREAD

**Issue by sherlock-admin3** [sherlock-audit/2024-03-zivoe-judging#486]
rbserver

medium

# Calling `ZivoeITO.migrateDeposits`, `ZivoeRewards.getRewards`, and `ZivoeRewardsVesting.getRewards` functions can revert if `transfer` function of one of relevant tokens fails due to contract pausing or sender/receiver blocking

## Summary
Calling the `ZivoeITO.migrateDeposits`, `ZivoeRewards.getRewards`, and `ZivoeRewardsVesting.getRewards` functions can revert if the `transfer` function of one of the relevant tokens fails due to contract pausing or sender/receiver blocking. Besides that this is a risk of external contracts pausing, which is unacceptable by Zivoe protocol, the relevant token amounts can be locked in the corresponding contracts.

## Vulnerability Detail
The `ZivoeITO.migrateDeposits`, `ZivoeRewards.getRewards`, and `ZivoeRewardsVesting.getRewards` functions in the Code Snippet section iterate over the corresponding tokens and transfer the relevant amount of each token to the target receiver. However, the `transfer` function calls of the tokens like USDC and USDT can revert when these tokens' contracts are paused or the corresponding sender and/or receiver are/is blocked by these tokens' admins according to these tokens' `transfer` functions shown in Code Snippet section. When the `transfer` function of one of the corresponding tokens stops working while the other tokens' `transfer` functions continue working fine, calling the `ZivoeITO.migrateDeposits`, `ZivoeRewards.getRewards`, and `ZivoeRewardsVesting.getRewards` functions revert because all of these tokens' `transfer` functions must be called due to these functions' loops.

## Impact
As a result, although only one token's `transfer` function stops working, the relevant amounts of the other tokens whose `transfer` functions are working fine cannot be transferred to the target receiver. If that one token's `transfer` function stops working indefinitely, these relevant token amounts will forever be locked in the corresponding contracts. This is a risk of external contracts pausing, which is not acceptable by the Zivoe protocol as stated in https://audits.sherlock.xyz/contests/280.

## Code Snippet
https://github.com/sherlock-audit/2024-03-zivoe/blob/main/zivoe-core-foundry/src/ZivoeITO.sol#L313-L335
```solidity
    function migrateDeposits() external {
        require(end != 0, "ZivoeITO::migrateDeposits() end == 0");
        if (_msgSender() != IZivoeGlobals_ITO(GBL).ZVL()) {
            require(block.timestamp > end, "ZivoeITO::migrateDeposits() block.timestamp <= end");
        }
        ...

        for (uint256 i = 0; i < stables.length; i++) {
            IERC20(stables[i]).safeTransfer(IZivoeGlobals_ITO(GBL).DAO(), IERC20(stables[i]).balanceOf(address(this)));
        }

        ...
    }
```

https://github.com/sherlock-audit/2024-03-zivoe/blob/main/zivoe-core-foundry/src/ZivoeRewards.sol#L281-L295
```solidity
    function getRewards() public updateReward(_msgSender()) {
        for (uint256 i = 0; i < rewardTokens.length; i++) { _getRewardAt(i); }
    }
    
    ...
    function _getRewardAt(uint256 index) internal nonReentrant {
        address _rewardsToken = rewardTokens[index];
        uint256 reward = rewards[_msgSender()][_rewardsToken];
        if (reward > 0) {
            rewards[_msgSender()][_rewardsToken] = 0;
            IERC20(_rewardsToken).safeTransfer(_msgSender(), reward);
            emit RewardDistributed(_msgSender(), _rewardsToken, reward);
        }
    }
```

https://github.com/sherlock-audit/2024-03-zivoe/blob/main/zivoe-core-foundry/src/ZivoeRewardsVesting.sol#L484-L498
```solidity
    function getRewards() public updateReward(_msgSender()) {
        for (uint256 i = 0; i < rewardTokens.length; i++) { _getRewardAt(i); }
    }
    
    ...
    function _getRewardAt(uint256 index) internal nonReentrant {
        address _rewardsToken = rewardTokens[index];
        uint256 reward = rewards[_msgSender()][_rewardsToken];
        if (reward > 0) {
            rewards[_msgSender()][_rewardsToken] = 0;
            IERC20(_rewardsToken).safeTransfer(_msgSender(), reward);
            emit RewardDistributed(_msgSender(), _rewardsToken, reward);
        }
    }
```

USDC's `transfer` function is as follows.
```solidity
    function transfer(address to, uint256 value)
        external
        override
        whenNotPaused
        notBlacklisted(msg.sender)
        notBlacklisted(to)
        returns (bool)

    {
        _transfer(msg.sender, to, value);
        return true;
    }
```

USDT's `transfer` function is as follows.
```solidity
    function transfer(address _to, uint _value) public whenNotPaused {
        require(!isBlackListed[msg.sender]);
        if (deprecated) {
            return UpgradedStandardToken(upgradedAddress).transferByLegacy(msg.sender, _to, _value);
        } else {
            return super.transfer(_to, _value);
        }
    }
```

## Tool used
Manual Review

## Recommendation
The `ZivoeITO.migrateDeposits`, `ZivoeRewards.getRewards`, and `ZivoeRewardsVesting.getRewards` functions can be respectively updated to add an input parameter for specifying the token(s) to be transferred and transfer such specified token(s) instead of iterating over and transferring all of the corresponding tokens at once.

Duplicate of #657

**Comment by pseudonaut** [comment]
Intended, not of concern

**Comment by rbserver** [comment]
Hi @Evert0x, thanks for your work! I noticed that this issue is still a duplicate of 13. Would this issue be updated to be a duplicate of https://github.com/sherlock-audit/2024-03-zivoe-judging/issues/657 based on @WangSecurity's https://github.com/sherlock-audit/2024-03-zivoe-judging/issues/13#issuecomment-2118442277?

**Comment by WangSecurity** [comment]
@rbserver should be correct now, please check

**Comment by rbserver** [comment]
> @rbserver should be correct now, please check

It looks correct now. Thanks for your help! @WangSecurity
